edx_rest_framework_extensions/permissions.py

```diff
@@ -83,21 +90,61 @@ def has_permission(self, request, view):
         of the organizations specified in the content provider filters
         in the JWT used to authenticate.
         """
-        if _should_skip_jwt_scope_enforcement(request):
-            return True
         course_key = CourseKey.from_string(view.kwargs.get('course_id'))
         jwt_filters = decode_jwt_filters(request.auth)
-        for provider_type, filter_value in jwt_filters:
-            if provider_type == 'content_org' and filter_value == course_key.org:
+        for filter_type, filter_value in jwt_filters:
+            if filter_type == 'content_org' and filter_value == course_key.org:
                 return True
+        log.warning("Permission JwtHasContentOrgFilterForRequestedCourse not satisfied.")
         return False
 
 
-def _should_skip_jwt_scope_enforcement(request):
+class JwtHasUserFilterForRequestedUser(BasePermission):
     """
-    Returns True if either the request is not JWT authenticated or if
-    JWT scopes enforcement is disabled.
+    The JWT used to authenticate contains the appropriate user filter for the
+    requested user resource.
     """
-    if not is_jwt_authenticated(request):
-        return True
-    return not waffle.switch_is_active(SWITCH_ENFORCE_JWT_SCOPES)
+    message = 'JWT missing required user filter.'
+
+    def has_permission(self, request, view):
+        """
+        If the JWT has a user filter, verify that the filtered
+        user value matches the user in the URL.
+        """
+        user_filter = self._get_user_filter(request)
+        if not user_filter:
+            return True
+
+        allowed = user_filter == _get_username_param(request)
+        if not allowed:
+            log.warning("Permission JwtHasUserFilterForRequestedUser not satisfied for user_filter %s.", user_filter)
+        return allowed
+
+    def _get_user_filter(self, request):
+        jwt_filters = decode_jwt_filters(request.auth)
+        for filter_type, filter_value in jwt_filters:
+            if filter_type == 'user':
+                if filter_value == 'me':
+                    filter_value = request.user.username.lower()
+                return filter_value
+        return None
+
+
+def jwt_course_access_or_is_user_in_url_or_staff():
+    not_jwt_restricted_permissions = C(NotJwtRestrictedApplication) & (C(IsStaff) | IsUserInUrl)
+    jwt_restricted_permissions = (
+        C(JwtRestrictedApplication) &
+        JwtHasScope &
+        JwtHasContentOrgFilterForRequestedCourse &
+        JwtHasUserFilterForRequestedUser
+    )
+    return (not_jwt_restricted_permissions | jwt_restricted_permissions)
```

Comment:
1. Why have this be a function instead of another importable conditional permission?
2. Is there something else that we could name it that better expresses the intent of putting all these permissions together? Or maybe break it up a bit more so that the jwt_restricted_permissions are one condition and we put them together in the view/view-decorator code? I think it gets confusing when there are two "or" in the function name and you don't know where the parens are supposed to be.

Reply:
1. It would be good to have it be a conditional permission.  But the middleware check doesn't find the required permission if I include it within a new permission class.  Would it work if I inherit from `Conditional` directly?
2. Oh yes PLEASE.  I named this late last night.  I beg for a better name than this.

Reply:
1. Weird. I thought I fixed the middleware check to drill into all the conditional permissions and pass if the target permission class is in at least one of them. Could you write out the `permission_classes` value that the middleware isn't searching through correctly? I can try debugging it on my side.

2. Yeah, that is a tough one. Is this the sort of standard set of permissions we're expecting to put on a legacy endpoint for compatibility? Are there other common variations that will be used?

Reply:
1. The middleware does work as designed and drills in while the Permission class is a Conditional.  But, if I wanted to do the following:

```
class CourseUserAccessPermission(BasePermission):
    def has_permission(self, request, view):
        not_jwt_restricted_permissions = C(NotJwtRestrictedApplication) & (C(IsStaff) | IsUserInUrl)
        jwt_restricted_permissions = (
            C(JwtRestrictedApplication) &
            JwtHasScope &
            JwtHasContentOrgFilterForRequestedCourse &
            JwtHasUserFilterForRequestedUser
        )
        return (not_jwt_restricted_permissions | jwt_restricted_permissions)
```

then it wouldn't catch that the implementation of the `Permission` class checks the required permission within it.

Reply:
2. What about:
```
jwt_restricted_application_or_user_access
```